cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -0,0 +1,379 @@
+      try (PrintWriter writer = new PrintWriter(startFile.getOutputStream())) {
+        writer.write(requestJson);
+      }
+      LOG.debug("Wrote to startFile {}", startFile.toURI());
+      // Generate the report asynchronously in a new thread
+      Executors.newSingleThreadExecutor().submit(() -> {
+        try {
+          generateReport(reportRequest, reportBaseDir);
+        } catch (Throwable t) {
+          LOG.error("Failed to generate report {}", reportId, t);
+          try {
+            // Write to the failure file in case of any exception occurs during report generation
+            Location failureFile = reportBaseDir.append(FAILURE_FILE);
+            failureFile.createNew();
+            try (PrintWriter writer = new PrintWriter(failureFile.getOutputStream())) {
+              writer.println(t.toString());
+              t.printStackTrace(writer);
+            }
+          } catch (Throwable t2) {
+            LOG.error("Failed to write cause of failure to file for report {}", reportId, t2);
+            throw new RuntimeException("Failed to write cause of failure to file for report " + reportId, t2);
+          }
+        }
+      });
+      responder.sendJson(200, ImmutableMap.of("id", reportId));
+    }
+
+    /**
+     * Generates report files according to the given request and write them to the given location.
+     * Program run meta files are first filtered to exclude unnecessary files for report generation,
+     * and send the paths of qualified run meta files to {@link ReportGenerationHelper#generateReport}
+     * that actually launches a Spark job to generate reports.
+     *
+     * @param reportRequest request
+     * @param reportBaseDir location of the directory where the report files directory, COUNT file,
+     *                      and _SUCCESS file will be created.
+     */
+    private void generateReport(ReportGenerationRequest reportRequest, Location reportBaseDir) throws IOException {
+      Location baseLocation = Transactionals.execute(getContext(), context -> {
+        return context.<FileSet>getDataset(ReportGenerationApp.RUN_META_FILESET).getBaseLocation();
+      });
+      // Get a list of directories of all namespaces under RunMetaFileset base location
+      List<Location> nsLocations;
+      nsLocations = baseLocation.list();
+      // Get the namespace filter from the request if it exists
+      ReportGenerationRequest.ValueFilter<String> namespaceFilter = null;
+      if (reportRequest.getFilters() != null) {
+        for (ReportGenerationRequest.Filter filter : reportRequest.getFilters()) {
+          if (ReportField.NAMESPACE.getFieldName().equals(filter.getFieldName())) {
+            // ReportGenerationRequest is validated to contain only one filter for namespace field
+            namespaceFilter = (ReportGenerationRequest.ValueFilter<String>) filter;
+            LOG.debug("Found namespace filter {}", namespaceFilter);
+            break;
+          }
```

Comment:
can't this logic be a method `getNamespaceFilterIfExists` that will return the namespace filter if its part of filter.